designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ScenarioLiveDataApiHttpService.scala

```diff
@@ -64,10 +69,55 @@ class ScenarioLiveDataApiHttpService(
                 Future.successful(Left(LiveDataNotSupported))
             }
           }
-        } yield LiveDataDto.from(liveData)
+        } yield LiveDataDto.from(liveData, computeCounts(scenarioWithDetails, liveData))
       }
   }
 
+  private def computeCounts(scenarioWithDetails: ScenarioWithDetails, liveData: LiveData)(
+      implicit loggedUser: LoggedUser
+  ): Map[String, NodeCount] = {
+    val uniqueNodeIds: Set[String] = liveData.nodeTransitions.keys.flatMap { t =>
+      t.destinationNodeId match {
+        case Some(dest) => List(t.sourceNodeId, dest)
+        case None       => List(t.sourceNodeId)
+      }
+    }.toSet
+
+    val incomingCounts =
+      liveData.nodeTransitions.toList
+        .flatMap { case (transition, data) => transition.destinationNodeId.map(_ -> data.totalCount) }
+        .groupBy(_._1)
+        .map { case (nodeId, counts) => nodeId -> counts.map(_._2).sum }
+
+    val outgoingCounts =
+      liveData.nodeTransitions.toList
+        .map { case (transition, data) => transition.sourceNodeId -> data.totalCount }
+        .groupBy(_._1)
+        .map { case (nodeId, counts) => nodeId -> counts.map(_._2).sum }
+
+    // We calculate counts based on transitions incoming to node, but if there are none (for sources) we use outgoing counts
+    val nodeCounts = uniqueNodeIds.map { nodeId =>
+      nodeId -> incomingCounts
+        .get(nodeId)
+        .orElse(outgoingCounts.get(nodeId))
+        .getOrElse(throw new IllegalStateException(s"Could not calculate counts for node $nodeId"))
+    }
+
+    val counts = nodeCounts.map { case (nodeId, count) =>
```

Comment:
Use can use `mapValuesNow` and shorten it a bit. BTW I see that after all we pass this map as a fuction so maybe instead of creating a `Map` we can just make a local function that would compute it ad-hoc. WDYT?

Reply:
Refactored as a function

Reply:
I mean sth like this:
```
    def getCount(nodeId: String): Option[RawCount] = {
      Some(
        RawCount(
          incomingCounts
            .get(nodeId)
            .orElse(outgoingCounts.get(nodeId))
            .getOrElse(0),
          liveData.exceptions.getOrElse(NodeId(nodeId), List.empty).size.toLong
        )
      )
    }
```